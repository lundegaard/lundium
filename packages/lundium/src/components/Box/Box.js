import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import { compose, filter, fromPairs, map, omit, test, toPairs } from 'ramda';
import { cx, isNilOrEmptyString } from 'ramda-extension';

import classNamesByBreakpoint from '../../utils/classNamesByBreakpoint';
import responsiveStringOrNumberPropType from '../../utils/types/responsiveStringOrNumberPropType';

const propToClassName = prefix =>
	classNamesByBreakpoint(
		(value, breakpoint) =>
			!isNilOrEmptyString(value) &&
			(breakpoint === 'xs'
				? `${prefix}-${value}`
				: `${prefix}-${breakpoint}-${value}`),
	);
const displayToClassName = propToClassName('d');

const omitInvalidProps = compose(
	fromPairs,
	// filter only event handlers and data-attributes.
	// Due to backwards compatibily we can't use `elementProps`
	filter(([prop]) => test(/^(on|data-|role|aria-)/, prop)),
	toPairs,
);

const DIMENSIONS = [
	'm',
	'p',
	'mx',
	'my',
	'mt',
	'mr',
	'mb',
	'ml',
	'px',
	'py',
	'pt',
	'pr',
	'pb',
	'pl',
	'order',
];

const Box = forwardRef(
	(
		{
			children,
			as: Comp = 'div',
			className,
			display,
			hasStrictAttributes,
			elementProps,
			...rest
		},
		ref,
	) => (
		<Comp
			className={cx(
				...map(x => propToClassName(x)(rest[x]))(DIMENSIONS),
				displayToClassName(display),
				className,
			)}
			ref={ref}
			{...(hasStrictAttributes
				? omitInvalidProps(rest)
				: omit(DIMENSIONS, rest))}
			{...elementProps}
		>
			{children}
		</Comp>
	),
);

Box.displayName = 'forwardRef(Box)';

const dimensionsPropTypes = DIMENSIONS.reduce(
	(propTypes, dimension) => ({
		...propTypes,
		[dimension]: responsiveStringOrNumberPropType,
	}),
	{},
);

Box.propTypes = {
	/** Tag used for root component */
	as: PropTypes.any,
	/** Children to be rendered in the main container. */
	children: PropTypes.node,
	/** Class name for root. */
	className: PropTypes.string,
	display: responsiveStringOrNumberPropType,
	elementProps: PropTypes.object,
	hasStrictAttributes: PropTypes.bool,
	...dimensionsPropTypes,
};

export default Box;
