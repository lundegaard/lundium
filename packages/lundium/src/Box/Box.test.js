import React from 'react';
import { shallow } from 'enzyme';

import Box from './Box';

const stringify = value => JSON.stringify(value, null, 2);

describe('<Box />', () => {
	const requiredProps = {};

	it('shallow', () => {
		const wrapper = shallow(<Box {...requiredProps} />);
		expect(wrapper).toMatchSnapshot();
	});

	describe('generate className correctly', () => {
		const testClassNames = (propsObject, expectedClassName) =>
			it(`for props equal to ${stringify(propsObject)}
			, generates className equal to "${expectedClassName}"`, () => {
				const wrapper = shallow(<Box {...propsObject} />);
				expect(wrapper.props().className).toBe(expectedClassName);
			});

		describe('for property m', () => {
			testClassNames({ m: [1, 2, 3, 4, 5] }, 'm-1 m-sm-2 m-md-3 m-lg-4 m-xl-5');
			testClassNames({ m: [1, 2, 3, 4] }, 'm-1 m-sm-2 m-md-3 m-lg-4');
			testClassNames({ m: [1, 2, 3] }, 'm-1 m-sm-2 m-md-3');
			testClassNames({ m: [1, 2] }, 'm-1 m-sm-2');
			testClassNames({ m: [1] }, 'm-1');
			testClassNames({ m: [] }, '');
			testClassNames(
				{ m: { xs: 1, sm: 2, md: 3, lg: 4, xl: 5 } },
				'm-1 m-sm-2 m-md-3 m-lg-4 m-xl-5',
			);
			testClassNames(
				{ m: { xs: 1, sm: 2, md: 3, lg: 4 } },
				'm-1 m-sm-2 m-md-3 m-lg-4',
			);
			testClassNames({ m: { xs: 1, sm: 2, md: 3 } }, 'm-1 m-sm-2 m-md-3');
			testClassNames({ m: { xs: 1, sm: 2 } }, 'm-1 m-sm-2');
			testClassNames({ m: { xs: 1 } }, 'm-1');
			testClassNames({ m: {} }, '');
		});

		describe('even for falsy values', () => {
			testClassNames({ m: { xs: 0, lg: 0 } }, 'm-0 m-lg-0');
		});
	});
});
