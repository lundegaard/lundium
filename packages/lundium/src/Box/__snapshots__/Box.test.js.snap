// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> shallow 1`] = `
<div
  className=""
/>
`;
